src/java/org/apache/cassandra/db/memtable/Memtable.java

```diff
@@ -0,0 +1,389 @@
+import org.apache.cassandra.utils.FBUtilities;
+import org.apache.cassandra.utils.concurrent.OpOrder;
+import org.apache.cassandra.utils.memory.MemtablePool;
+import org.psjava.ds.Collection;
+
+/**
+ * Memtable interface. This defines the operations the ColumnFamilyStore can perform with memtables.
+ * They are of several types:
+ * - construction factory interface
+ * - write and read operations: put, getPartition and makePartitionIterator
+ * - statistics and features, including partition counts, data size, encoding stats, written columns
+ * - memory usage tracking, including methods of retrieval and of adding extra allocated space (used non-CFS secondary
+ *   indexes)
+ * - flush functionality, preparing the set of partitions to flush for given ranges
+ * - lifecycle management, i.e. operations that prepare and execute switch to a different memtable, together
+ *   with ways of tracking the affected commit log spans
+ */
+public interface Memtable extends Comparable<Memtable>
+{
+    // Construction
+
+    /**
+     * Factory interface for constructing memtables, and querying write durability features.
+     *
+     * The factory is chosen using the MemtableParams class (passed as argument to
+     * {@code CREATE TABLE ... WITH memtable = {...}} or in the memtable options in cassandra.yaml). To make that
+     * possible, implementations must provide either a static {@code FACTORY} field (if they accept no further option)
+     * or a static {@code factory(Map<String, String>)} method. In the latter case, the method should avoid creating
+     * multiple instances of the factory for the same parameters, or factories should at least implement hashCode and
+     * equals.
+     */
+    interface Factory
+    {
+        /**
+         * Create a memtable.
+         *
+         * @param commitLogLowerBound A commit log lower bound for the new memtable. This will be equal to the previous
+         *                            memtable's upper bound and defines the span of positions that any flushed sstable
+         *                            will cover.
+         * @param metadaRef Pointer to the up-to-date table metadata.
+         * @param owner Owning objects that will receive flush requests triggered by the memtable (e.g. on expiration).
+         */
+        Memtable create(AtomicReference<CommitLogPosition> commitLogLowerBound, TableMetadataRef metadaRef, Owner owner);
+
+        /**
+         * If the memtable can achieve write durability directly (i.e. using some feature other than the commitlog, e.g.
+         * persistent memory), it can return true here, in which case the commit log will not store mutations in this
+         * table.
+         * Note that doing so will prevent point-in-time restores and changed data capture, thus a durable memtable must
+         * allow the option of turning commit log writing on even if it does not need it.
+         */
+        default boolean writesShouldSkipCommitLog()
```

Comment:
I don't understand why should a memtable make a decision whether something goes into the commit log. It is not it's responsibility.
I can perfectly see the same Factory creating the same Memtable regardless of whether stuff should or should not go into the commit log.
It is actually explained in the comment - so why have this interface method at all? Why entangle Memtable with separate feature like CDC or point-in-time restores? What about any new features that would force commit log usage?

It seems natural to have logic that decides whether to put a write in the commit log or not on basis of `writesAreDurable` and enabled features. In the commit log or in CFS perhaps.

Reply:
The explanation is that I tend to look at this from the point of view of the local storage system, where the memtable and commitlog are actually two very tightly bound parts of the first component of the LSM tree. In that respect, I see the memtable as something that definitely needs to have full control over the commit log -- we have some of that already in the flush (by the log) and the recycling (by the memtable) control -- and even that it should logically contain it. There are also good benefits to be had in further integrating commit log with memtable e.g. in not duplicating data in the memtable when it's stored in the commit log. This also means that I kind of see CDC and point-in-time as not really belonging with the (local node) commit log either.

However, we do have them and they have to be accommodated somehow, and I agree that this configuration (whether or not to store already-persistent data for CDC or PITR) makes better sense together with the `durable_writes` and `cdc` keyspace options. I will move it there.

Reply:
My understanding is as follows:
There are two crucial properties of a write:
- persistence
- readability
Having both is necessary (or is it?) to confirm a write.

Typicaly (let's put persistent memtables away) sstables provide both, commit log provides persistence (but not ad-hoc readability) and memtable provides readability (but not persistence).
Then, typical memtable and commitlog are integrated together to provide persistence and readability before sstable is flushed.
Should I be rather thinking that we use the name "memtable" for the integration of the commitlog and the in-memory component?

Were it not for CDC or point-in-time the commit log would be an implementation detail of a memtable then? A detail that provides perstistence? 
 
 OK, so adding persistent memtable to the mix what we want is:
 - commit log as-is (CDC etc)
 - the in-memory component that provides persistence guarantees 
 - the integrating code which can now:
    - confirm writes when the write is acked by the in-memory component
    - trigger commitlog on or off if the in-memory component provides persistence and some feature requiring commitlog is enabled (it's an optimization)

Is the above correct?

> There are also good benefits to be had in further integrating commit log with memtable e.g. in not duplicating data in the memtable when it's stored in the commit log.

Could you please explain? What about readability? Will the commit log support reading?

Reply:
> Should I be rather thinking that we use the name "memtable" for the integration of the commitlog and the in-memory component?
> Were it not for CDC or point-in-time the commit log would be an implementation detail of a memtable then? A detail that provides perstistence?

Yes, this is the point of view of this API/pull request and the rationale behind the decision to steer Intel's pmem solution to a memtable one.

> Could you please explain? What about readability? Will the commit log support reading?

The idea there is for the memtable to be an index into the commitlog, which is kept partially in memory. This similar in design to an "append-only" structure, with an efficient in-memory structure for the parts that need constant updating. See [here](https://docs.google.com/presentation/d/1d6yxO_J8Rnph6E0F_UpTuEWdnrKBOvnE2RWgmp1TNW8/edit#slide=id.g3366841456_0_159) for details.

Reply:
I started to move the forced durability option, but it's turning out to be more invasive than I expected and not improving clarity. It looks like it is more natural to configure this if `skipCommitLog` remains in the memtable options, especially as this can avoid exposing configuration options elsewhere that are only useful for persistent memtables.

Do you insist on taking it out of the memtable options?

Reply:
Let's leave it as is then